### DISCUSSION TITLE
[QUESTION] amazon profile environment variable

### DISCUSSION THREAD

**Discussion by thelittlerocket** [le0pard/certonid#8] [Q&A] [answered]
Hi there,
Thanks for the hard work and after setting things up, the script is now working.
Currently, I am using AWS profile to control which ID can request to sign their key, but this will require each user to have aws cli installed. 
So, would like to know if want to use the environment variable, how to set up for multiple user accounts on Lambda?

Thanks,

**Comment by le0pard** [comment, marked as answer]
Hello @thelittlerocket 

Can you provide more info about issue? Because 

"I am using AWS profile to control which ID can request to sign their key, but this will require each user to have aws cli installed"

 and 

"would like to know if want to use the environment variable, how to set up for multiple user accounts on Lambda" 

looks like different issues with different components - cli and serverless. Or I am not fully understand issue.

**Reply by thelittlerocket** [reply]
Hi @le0pard 
Thanks for replying and sorry for not describing my question clearly.

I was following this guide https://github.com/certonid/certonid/wiki/Getting-Started , 
using the symmetric method to encrypt my CA key password and CA key,
then using the result of randstr as the key of an environment variable CERTONID_SYMMETRIC_KEY on Lambda's setting page
https://share.getcloudapp.com/8Lun9jPv

then I upload all components to Lambda, using "serverless" as handler which is the file name of the serverless binary
https://share.getcloudapp.com/2Nu0PX0m

serverless certonid.yml
`
ca:
  storage: file
  path: ca.enc
  encrypted:
    encryption: symmetric
  passphrase:
    encryption: symmetric
    content: "result_from_randstr"
certificates:
  user:
    max_valid_until: 2h
    additional_principals:
      - "ubuntu"
    critical_options:
      - "source-address 0.0.0.0/0"
    extensions:
    - "permit-X11-forwarding"
    - "permit-agent-forwarding"
    - "permit-port-forwarding"
    - "permit-pty"
    - "permit-user-rc"
  host:
    max_valid_until: 2h
    additional_principals:
      - “*.domain.tld”
    critical_options:
      - "source-address 0.0.0.0/0"
    extensions:
      - "permit-X11-forwarding"
      - "permit-agent-forwarding"
      - "permit-port-forwarding"
      - "permit-pty"
      - "permit-user-rc"

logger:
  level: debug
  format: text
`
the client-side( requesting their key to be signed so they can log in to the server), I follow this guide
https://github.com/certonid/certonid/wiki/Certonid-CLI-config
and on the client-side, has aws cli installed and have two profile configured
 .certonid.yml

cache_path: ~/.certonid_cache
certificates:
  server1:
    public_key_path: ~/.ssh/server1.pub
    username: root
    runner: aws
    valid_until: 30m
    aws:
      profile: ssh-ca-macbookpro
      region: us-east-1
      function_name: serverless
    failover:
    - region: us-east-1
    - region: us-east-1
      function_name: serverless

  server2:
    public_key_path: ~/.ssh/server2.pub
    username: root
    runner: aws
    valid_until: 1h
    aws:
      profile: ssh-ca-macmini
      region: us-east-1
      function_name: serverless
logger:
  level: debug
  format: text

============

this is my configuration and it is working, I can request my own key to be signed and log in to the remote server without a problem,
I was wondering if there is a way that I can control which user can request their key to be signed without using aws profile, but as you replied that I can create the relevant files without need aws cli to be installed, then this part is fine for me. Thanks.

and now I got another question, is it possible to have multiple CA in the same serverless configure or I have to use a different serverless configuration with another Lambda function?

Thanks,

**Reply by le0pard** [reply]
Thanks @thelittlerocket 

Now it is clear

> I was wondering if there is a way that I can control which user can request their key to be signed without using aws profile, but as you replied that I can create the relevant files without need aws cli to be installed, then this part is fine for me. Thanks.

aws profiles need for people, which have different certonid aws functions for different projects (each project have own access by CA). If user need access only to one project, it can just use AWS env variables with certonid cli to call single AWS lambda. You allow or deny access to call AWS lambda by adding or removing user to IAM group, which allow to "call/invoke" AWS Lambda function.

> and now I got another question, is it possible to have multiple CA in the same serverless configure or I have to use a different serverless configuration with another Lambda function?

One CA = separate AWS lambda function. You can create multiple AWS lambdas with different CA files and manage different AWS IAM groups, which have access to each function. Only people, who will be in a group to needed function will be able to invoke AWS lambda function with certonid. 

You can see this workflow in terraform module - it create separate group for clients and allow this group to invoke AWS certonid lambda:
https://github.com/certonid/terraform-aws-certonid/blob/master/main.tf#L145-L172

After this it add defined users to this group, so they can use own AWS credentials to call this function from certonid cli: https://github.com/certonid/terraform-aws-certonid/blob/master/main.tf#L216-L222

If user need to be removed from access to get new certs for CA - you just remove it from AWS group and it will stop get new certificates (old one will expire).

Hope this information help.

**Reply by M96268004** [reply]
Thanks, @le0pard, very helpful information, thanks.

**Comment by le0pard** [comment]
BTW, did you setup it by using terraform module ( https://github.com/certonid/terraform-aws-certonid ) or use some other way?

**Comment by le0pard** [comment]
BTW, aws cli credentials is store in plain files:

> Where you find your home directory location varies based on the operating system, but is referred to using the environment variables %UserProfile% in Windows and $HOME or ~ (tilde) in Unix-based systems. You can specify a non-default location for the config file by setting the AWS_CONFIG_FILE environment variable to another local path

> For example, the files generated by the CLI for a default profile configured with aws configure looks similar to the following.

~/.aws/credentials
```
[default]
aws_access_key_id=AKIAIOSFODNN7EXAMPLE
aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
[something-else]
aws_access_key_id=AKIAIOSFODNN7EXAMPLE
aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```

~/.aws/config
```
[default]
region=us-west-2
[something-else]
region=us-east-1
```

So you can create and store any amount of aws profiles without having aws cli installed on machine @thelittlerocket
